### PR TITLE
Modify hypertable catalog update API

### PR DESCRIPTION
**Pull request by gayyappan**
For reviewers:
multiple commits to help with reviewing. I'll rebase as 1 commit after reviews.

Disable-check: force-changelog-file

### CODE REVIEW

**Comment by mkindahl** [comment]
> For reviewers: multiple commits to help with reviewing. I'll rebase as 1 commit after reviews.
> 
> Disable-check: force-changelog-file

Having a rationale (a "why" is this change made) in the PR would help with reviewing.

**Review by mkindahl — COMMENTED** [review]
Looks functionally good, but some questions:

- Not sure about some changes to use `ts_hypertable_update_status_osm` rather than `ts_hypertable_update`. Was this wrong before or is it a change in behavior?
- The set functions returns true always even though the return value is not checked in any of the call sites. Might be better to change the function signature.
- Metadata corruption (or bugs where the row is removed before you have a chance to lock it) might result in tuples not being found. If this happens, it seems like execution should not proceed so it makes more sense to throw an error in release builds.
- You have one check if the updated value is new but no check in the other cases. It might be prudent to be consistent.
- I am not sure if a new row is generated for each update even if there are no changes, so checking for equality might avoid generating dead rows when the values are equal.

File: src/chunk.c [hunk 1]
```diff
@@ -4707,7 +4707,7 @@ add_foreign_table_as_chunk(Oid relid, Hypertable *parent_ht)
 	parent_ht->fd.status =
 		ts_set_flags_32(parent_ht->fd.status,
 						HYPERTABLE_STATUS_OSM | HYPERTABLE_STATUS_OSM_CHUNK_NONCONTIGUOUS);
-	ts_hypertable_update(parent_ht);
+	ts_hypertable_update_status_osm(parent_ht);
```
[the remark this answers is not in the document]
Comment (gayyappan) [reply src/chunk.c:4710]: You can no longer execute a blanket ts_hypertable_update call. Every field has a filed specific API. You can choose to blindly overwrite the field (as in the case of name and schema)  or do something different after checking the current value of the tuple (like the status check at line 2649).

File: src/hypertable.c [hunk 2]
```diff
@@ -403,59 +403,88 @@ hypertable_scan_limit_internal(ScanKeyData *scankey, int num_scankeys, int index
 	return ts_scanner_scan(&scanctx);
 }
 
-static ScanTupleResult
-hypertable_tuple_update(TupleInfo *ti, void *data)
+// update the tuple at this tid. The assumption is that we already hold a
+// tuple exclusive lock and no other transaction can modify this tuple
```
[the remark this answers is not in the document]
Comment (gayyappan) [reply src/hypertable.c:407]: fixed

**Comment by gayyappan** [comment]
> Looks functionally good, but some questions:
> 
> * Not sure about some changes to use `ts_hypertable_update_status_osm` rather than `ts_hypertable_update`. Was this wrong before or is it a change in behavior?
The update APIs are field specific now. No more blanket calls. We were not consistent with this.

> * The set functions returns true always even though the return value is not checked in any of the call sites. Might be better to change the function signature.
I didn't want to change the interface of these external functions . will trigger changes at all caller sites.

> * Metadata corruption (or bugs where the row is removed before you have a chance to lock it) might result in tuples not being found. If this happens, it seems like execution should not proceed so it makes more sense to throw an error in release builds.
Yes, you are right. replaced Assert  with Ensure.

> * You have one check if the updated value is new but no check in the other cases. It might be prudent to be consistent. I am not sure if a new row is generated for each update even if there are no changes, so checking for equality might avoid generating dead rows when the values are equal

This is to make sure that there is no change in the current behavior. Only the status field could have frequent updates since it is triggered by an internal call (via osm).

File: src/hypertable.c [hunk 3]
```diff
@@ -2397,38 +2456,52 @@ ts_hypertable_has_compression_table(const Hypertable *ht)
  * 1. RowShareLock to SELECT for UPDATE
  * 2. UPDATE status using RowExclusiveLock
  */
-static bool
-hypertable_update_status_osm(Hypertable *ht)
+bool
+ts_hypertable_update_status_osm(Hypertable *ht)
```
[the remark this answers is not in the document]
Comment (gayyappan) [reply src/hypertable.c:2460]: Yes, this is triggered only via OSM right now. See statuses here: https://github.com/timescale/timescaledb/blob/ea2284386b8e16268f51d60fe920b1815868d78d/src/ts_catalog/catalog.h#L1236

**Review by mkindahl — APPROVED** [review]

File: src/hypertable.c [hunk 4]
```diff
@@ -2397,38 +2463,52 @@ ts_hypertable_has_compression_table(const Hypertable *ht)
...
-			success = ts_hypertable_update(ht); // get RowExclusiveLock and update here
-		}
+bool
+ts_hypertable_update_chunk_sizing(Hypertable *ht)
+{
+	FormData_hypertable form;
+	ItemPointerData tid;
+	/* lock the tuple entry in the catalog table */
+	bool found = lock_hypertable_tuple(ht->fd.id, &tid, &form);
+	Ensure(found, "hypertable id %d not found", ht->fd.id);
```
Comment (fabriziomello) [inline src/hypertable.c:2491]: Is this message correct?

Comment (gayyappan) [reply src/hypertable.c:2491]: Yes, locking failures have their error messages. So if we get false here, it means that we couldn't find the tuple.

File: src/hypertable.c [hunk 5]
```diff
@@ -403,59 +403,95 @@ hypertable_scan_limit_internal(ScanKeyData *scankey, int num_scankeys, int index
...
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INTERNAL_ERROR),
+						 errmsg("unable to lock hypertable catalog tuple, lock result is %d for "
+								"hypertable "
+								"ID (%d)",
+								ti->lockresult,
+								htid)));
+			}
+		}
```
Comment (fabriziomello) [inline src/hypertable.c:485]: Do we have tests for it?

Comment (gayyappan) [reply src/hypertable.c:485]: Happy to add one if you can think of a way of triggering this error.

**Comment by mkindahl** [comment]
> > Looks functionally good, but some questions:
> > 
> > * Not sure about some changes to use `ts_hypertable_update_status_osm` rather than `ts_hypertable_update`. Was this wrong before or is it a change in behavior?
> 
> The update APIs are field specific now. No more blanket calls. We were not consistent with this.

Could you add a comment about this so that nobody changes back to use the original function?

> > * The set functions returns true always even though the return value is not checked in any of the call sites. Might be better to change the function signature.
> 
> I didn't want to change the interface of these external functions . will trigger changes at all caller sites.

I checked a few call sites and they were not using the return value. However, this is a nit.

File: src/hypertable.c [hunk 6]
```diff
@@ -2232,10 +2291,17 @@ ts_hypertable_set_compressed(Hypertable *ht, int32 compressed_hypertable_id)
...
 {
+	FormData_hypertable form;
+	ItemPointerData tid;
+	/* lock the tuple entry in the catalog table */
+	bool found = lock_hypertable_tuple(ht->fd.id, &tid, &form);
+	Ensure(found, "hypertable id %d not found", ht->fd.id);
+
 	Assert(!TS_HYPERTABLE_IS_INTERNAL_COMPRESSION_TABLE(ht));
 	ht->fd.compression_state = HypertableCompressionOff;
 	ht->fd.compressed_hypertable_id = INVALID_HYPERTABLE_ID;
```
[the remark this answers is not in the document]
Comment (gayyappan) [reply src/hypertable.c:2302]: Thanks for figuring this one out.

**Pull request merged**